lib/core/test/unit/Network/Wai/Middleware/LoggingSpec.hs

```diff
@@ -0,0 +1,301 @@
+    )
+import Network.HTTP.Types.Header
+    ( hContentType )
+import Network.Socket
+    ( Socket )
+import Network.Wai.Handler.Warp
+    ( runSettingsSocket, setBeforeMainLoop )
+import Network.Wai.Middleware.Logging
+    ( ApiLoggerSettings, newApiLoggerSettings, obfuscateKeys, withApiLogger )
+import Servant
+    ( (:<|>) (..)
+    , (:>)
+    , Application
+    , DeleteNoContent
+    , Get
+    , JSON
+    , NoContent (..)
+    , OctetStream
+    , PostCreated
+    , ReqBody
+    , Server
+    , err400
+    , err500
+    , serve
+    , throwError
+    )
+import Servant.Server
+    ( Handler )
+import Test.Hspec
+    ( Spec, after, afterAll, beforeAll, describe, it, shouldBe, shouldContain )
+
+import qualified Data.Aeson as Aeson
+import qualified Data.Map.Strict as Map
+import qualified Data.Text as T
+import qualified Network.Wai.Handler.Warp as Warp
+
+spec :: Spec
+spec = describe "Logging Middleware"
+    $ beforeAll setup $ after clearLogs $ afterAll tearDown $ do
+    it "GET, 200, no query" $ \ctx -> do
+        get ctx "/get"
+        expectLogs ctx
+            [ (Info, "[GET] /get")
+            , (Info, "200 OK")
+            , (Debug, "14")
+            ]
+
+    it "GET, 200, with query" $ \ctx -> do
+        get ctx "/get?query=patate"
+        expectLogs ctx
+            [ (Info, "[GET] /get?query=patate")
+            , (Info, "200 OK")
+            , (Debug, "14")
+            ]
+
+    it "GET, 200, not json" $ \ctx -> do
+        get ctx "/not-json"
+        expectLogs ctx
+            [ (Info, "[GET] /not-json")
+            , (Info, "200 OK")
+            , (Debug, "\NUL\NUL\NUL")
+            ]
+
+    it "POST, 201, with sensitive fields" $ \ctx -> do
+        post ctx "/post" (MkJson { field = "patate", sensitive = 14 })
+        expectLogs ctx
+            [ (Info, "[POST] /post")
+            , (Debug, "{\"sensitive\":\"*****\",\"field\":\"patate\"}")
+            , (Info, "201 Created")
+            , (Debug, "{\"status\":\"ok\",\"whatever\":42}")
+            ]
+
+    it "POST, 400, invalid payload (not json)" $ \ctx -> do
+        postIlled ctx "/post" "\NUL\NUL\NUL"
+        expectLogs ctx
+            [ (Info, "[POST] /post")
+            , (Debug, "Invalid payload: not JSON")
+            , (Info, "400 Bad Request")
+            , (Debug, "Failed reading: not a valid json value")
+            ]
+
+    it "DELETE, 202, no query" $ \ctx -> do
+        delete ctx "/delete"
+        expectLogs ctx
+            [ (Info, "[DELETE] /delete")
+            , (Info, "204 No Content")
+            ]
+
+    it "GET, 400" $ \ctx -> do
+        get ctx "/error400"
+        expectLogs ctx
+            [ (Info, "[GET] /error400")
+            , (Info, "400 Bad Request")
+            ]
+
+    it "GET, 500" $ \ctx -> do
+        get ctx "/error500"
+        expectLogs ctx
+            [ (Info, "[GET] /error500")
+            , (Error, "500 Internal Server Error")
+            ]
+
+    it "different request ids" $ \ctx -> do
```

Comment:
we do not lose any request's log. but do we also have time order maintained in this setup? Maybe we do not have this kind of  guarantee here?

Reply:
Hmmm. Good point. There's actually no tests about the timing of request at all. It could be worth adding a handcrafted "long" endpoint (like, over a second or so) and check that the computed time is indeed over a certain amount of time.

For the concurrent test... 🤔.. Perhaps it'll be sufficient to also make a single request to that endpoint, and expect to have only one set of log lines with a time above 1s?

What do you think?

Reply:
(I mean,  single request to the long endpoint in addition to the 100 requests to the other faster one)

Reply:
exactly, how the logging will behave when timeout, also crossed my mind!

Reply:
Hmmm.. Timeout is only a client concern here. The server does not timeout so to speak, only the client closes the connection.

I am not actually 100% sure how servant / warp behaves in that case. I suppose they leave the handler until it completes and throwd when trying to write the response back to the client.. In any case, that's kinda out of scope for the middleware here which has no control over this.

Reply:
I was thinking that maybe we can add one property to those unit tests. Here we can (this would strengthen the case) and we would generate 100 **different endpoints** queries. And in the property we would realize that combination of request **twice** and compare results. With fast endpoints, then with long endpoints, or even better mix of them...Maybe it is overkill - but still it would give me more information what to expect from our logging engine.

Reply:
🤔 And what would we expect this property to be / demonstrate?

Reply:
that when we run the set of 100 different queries the logging is deterministic in terms of logging time order when something is satisfied? Just speculating ....

Reply:
Actually, having something deterministic won't be that easy because we don't know what servant and the GHC runtimes does with the requests. Eventually, Servant creates one thread per request handler, but they get interleaved in a non deterministic order so I wouldn't rely too much on that. So, we can't be sure that if req#1 is received before req#2, it'll log req#1 before. Especially since there's a MVar read before the log entry, it can cause the thread to hang and pass on the focus to another thread. Plus, I don't think it really matters to have logs in order for this kind of scenarios. What matters is that all the logs are indeed there.

So, best we can do is really say that, after `n` requests, there should be entries regarding `n` different requests.